### ISSUE TITLE
Segment all the objects of a video

### ISSUE THREAD

**Issue by Masrur02** [facebookresearch/segment-anything-2#105]
Hi, thanks for the amazing work. I understood that SAM2 is tracking single or multiple objects in a video. Is there any option to predict all the objects of a video without any prompting? If yes, how fast is the prediction?

**Comment by TaoshuaiZ** [comment]
The same question, could SAM2 segment all the objects(like SAM) and track them in a video?

**Comment by chenkangjie1123** [comment]
The same question

**Comment by danu9327** [comment]
the same question

I'm curious too

**Comment by 23wjj** [comment]
the same question

**Comment by YTEP-ZHI** [comment]
same question.

**Comment by heyoeyo** [comment]
> Is there any option to predict all the objects of a video without any prompting

It doesn't look like there is any example code for this specific use case, but as a starting point, you could combine the [auto-mask generation notebook code](https://github.com/facebookresearch/segment-anything-2/blob/main/notebooks/automatic_mask_generator_example.ipynb) with the [video segmentation notebook code](https://github.com/facebookresearch/segment-anything-2/blob/main/notebooks/video_predictor_example.ipynb) to do this. Something like:

```python
import os
from PIL import Image
import numpy as np
import torch
from sam2.build_sam import build_sam2, build_sam2_video_predictor
from sam2.automatic_mask_generator import SAM2AutomaticMaskGenerator

# Setup
video_path = "notebooks/videos/bedroom"
sam2_checkpoint = "checkpoints/sam2_hiera_tiny.pt"
model_cfg = "sam2_hiera_t.yaml"
device = "cuda"

# Auto-masking of first frame (from automatic mask generation notebook)
sam2 = build_sam2(model_cfg, sam2_checkpoint, device=device, apply_postprocessing=False)
first_frame_path = os.path.join(video_path, os.listdir(video_path)[0])
first_frame = Image.open(first_frame_path)
first_frame = np.array(first_frame.convert("RGB"))
mask_generator = SAM2AutomaticMaskGenerator(sam2)
auto_masks = mask_generator.generate(first_frame)
print("Number of auto-masks:", len(auto_masks))

# Add every 'auto-mask' as it's own prompt for video tracking
predictor = build_sam2_video_predictor(model_cfg, sam2_checkpoint, device=device)
inference_state = predictor.init_state(video_path=video_path)
dtype = next(predictor.parameters()).dtype
lowres_side_length = predictor.image_size // 4
for mask_idx, mask_result in enumerate(auto_masks):

    # Get mask into form expected by the model
    mask_tensor = torch.tensor(mask_result["segmentation"], dtype=dtype, device=device)
    lowres_mask = torch.nn.functional.interpolate(
        mask_tensor.unsqueeze(0).unsqueeze(0),
        size=(lowres_side_length, lowres_side_length),
        mode="bilinear",
        align_corners=False,
    ).squeeze()

    # Add each mask as it's own 'object' to segment
    _, out_obj_ids, out_mask_logits = predictor.add_new_mask(
        inference_state=inference_state,
        frame_idx=0,
        obj_id=mask_idx,
        mask=lowres_mask,
    )

# Do video segmentation (same as video segmentation notebook)
video_segments = {}
for out_frame_idx, out_obj_ids, out_mask_logits in predictor.propagate_in_video(inference_state):
    video_segments[out_frame_idx] = {
        out_obj_id: (out_mask_logits[i] > 0.0).cpu().numpy() for i, out_obj_id in enumerate(out_obj_ids)
    }
```

Since it's using masks as prompts, I'd expect the quality to be worse than if it's given points/boxes, so I'm sure there are ways this could be improved. For example, point prompts could be generated from the masks (which may work better?) or even better would be to use something like a YOLO model to provide the initial prompts, so you don't end up tracking uninteresting junk in the background.

**Comment by Masrur02** [comment]
Hi,

Thank you for the code.

Best,
Khan

**Comment by braxtonowens** [comment]
> Since it's using masks as prompts, I'd expect the quality to be worse than if it's given points/boxes, so I'm sure there are ways this could be improved. For example, point prompts could be generated from the masks (which may work better?) or even better would be to use something like a YOLO model to provide the initial prompts, so you don't end up tracking uninteresting junk in the background.

For SAM1 I did something similar where I saved the masks then sampled from them. To avoid sampling too close to the edges I used binary erosion on the masks.

**Comment by JINNMnm** [comment]
> > Is there any option to predict all the objects of a video without any prompting
> 
> It doesn't look like there is any example code for this specific use case, but as a starting point, you could combine the [auto-mask generation notebook code](https://github.com/facebookresearch/segment-anything-2/blob/main/notebooks/automatic_mask_generator_example.ipynb) with the [video segmentation notebook code](https://github.com/facebookresearch/segment-anything-2/blob/main/notebooks/video_predictor_example.ipynb) to do this. Something like:
> 
> ```python
> import os
> from PIL import Image
> import numpy as np
> import torch
> from sam2.build_sam import build_sam2, build_sam2_video_predictor
> from sam2.automatic_mask_generator import SAM2AutomaticMaskGenerator
> 
> # Setup
> video_path = "notebooks/videos/bedroom"
> sam2_checkpoint = "checkpoints/sam2_hiera_tiny.pt"
> model_cfg = "sam2_hiera_t.yaml"
> device = "cuda"
> 
> # Auto-masking of first frame (from automatic mask generation notebook)
> sam2 = build_sam2(model_cfg, sam2_checkpoint, device=device, apply_postprocessing=False)
> first_frame_path = os.path.join(video_path, os.listdir(video_path)[0])
> first_frame = Image.open(first_frame_path)
> first_frame = np.array(first_frame.convert("RGB"))
> mask_generator = SAM2AutomaticMaskGenerator(sam2)
> auto_masks = mask_generator.generate(first_frame)
> print("Number of auto-masks:", len(auto_masks))
> 
> # Add every 'auto-mask' as it's own prompt for video tracking
> predictor = build_sam2_video_predictor(model_cfg, sam2_checkpoint, device=device)
> inference_state = predictor.init_state(video_path=video_path)
> dtype = next(predictor.parameters()).dtype
> lowres_side_length = predictor.image_size // 4
> for mask_idx, mask_result in enumerate(auto_masks):
> 
>     # Get mask into form expected by the model
>     mask_tensor = torch.tensor(mask_result["segmentation"], dtype=dtype, device=device)
>     lowres_mask = torch.nn.functional.interpolate(
>         mask_tensor.unsqueeze(0).unsqueeze(0),
>         size=(lowres_side_length, lowres_side_length),
>         mode="bilinear",
>         align_corners=False,
>     ).squeeze()
> 
>     # Add each mask as it's own 'object' to segment
>     _, out_obj_ids, out_mask_logits = predictor.add_new_mask(
>         inference_state=inference_state,
>         frame_idx=0,
>         obj_id=mask_idx,
>         mask=lowres_mask,
>     )
> 
> # Do video segmentation (same as video segmentation notebook)
> video_segments = {}
> for out_frame_idx, out_obj_ids, out_mask_logits in predictor.propagate_in_video(inference_state):
>     video_segments[out_frame_idx] = {
>         out_obj_id: (out_mask_logits[i] > 0.0).cpu().numpy() for i, out_obj_id in enumerate(out_obj_ids)
>     }
> ```
> 
> Since it's using masks as prompts, I'd expect the quality to be worse than if it's given points/boxes, so I'm sure there are ways this could be improved. For example, point prompts could be generated from the masks (which may work better?) or even better would be to use something like a YOLO model to provide the initial prompts, so you don't end up tracking uninteresting junk in the background.

hey! thanks for your code.

I'm wondering, if there is some part that you don't see in the first frame, then the corresponding mask will not be generated. Such that when we propagate in video, that part will missing maybe?

Have you encounter such problem? If so, how can we solve it?

**Comment by heyoeyo** [comment]
> I'm wondering, if there is some part that you don't see in the first frame, then the corresponding mask will not be generated. Such that when we propagate in video, that part will missing maybe? Have you encounter such problem? If so, how can we solve it?

If only the SAM model is being used, then I would guess the best option would be to run the auto-mask + video segmentation for only a few seconds at a time (can be done with the [max_frame_num_to_track](https://github.com/facebookresearch/segment-anything-2/blob/7e1596c0b6462eb1d1ba7e1492430fed95023598/sam2/sam2_video_predictor.py#L651) option) and then repeatedly re-run the auto-masking (+ video segmentation) to restart the tracking, in case new objects appear. It might also make sense to run the tracking forward and backwards (using the [reverse](https://github.com/facebookresearch/segment-anything-2/blob/7e1596c0b6462eb1d1ba7e1492430fed95023598/sam2/sam2_video_predictor.py#L652) flag) over a short time window if you want as complete tracking on objects as possible.

This would introduce an annoying issue of having to 'match' objects across each of the short tracking time spans (i.e. each time the tracking is reset, the objects would look unique, even if you already saw them earlier). However, I think it should be possible to match masks by comparing their IoU scores. So for example if an object with 'mask A' at the end of the first 5 seconds has a very high IoU with 'mask B' from the start of the next 5 seconds, then you could say the results are matching and treat them as a single continuous object being tracked.

**Comment by JINNMnm** [comment]
@heyoeyo 

Thanks for your advice! I'll give it a try and update my result later.

**Comment by zrporz** [comment]
https://github.com/zrporz/AutoSeg-SAM2
We provide an implementation script for automatic segmentation that finds and tracks new objects by detecting the iou of the current unsegmented part with a static mask, feel free to use it!

**Comment by JINNMnm** [comment]
Thank you!!! It helps a lot.

**Comment by yuchenlichuck** [comment]
same question ahahah

**Comment by matteo-tafuro** [comment]
I was building an application with similar requirements and I came across a paper that extends SAM2 for this purpose. If you need to segment both _things_ and _stuff_, [EntitySAM](https://github.com/ymq2017/entitysam) (CVPR2025) performs automatic segmentation and tracking of novel entities without requiring user specified prompts.

**Comment by chenkangjie1123** [comment]
> I was building an application with similar requirements and I came across a paper that extends SAM2 for this purpose. If you need to segment both _things_ and _stuff_, [EntitySAM](https://github.com/ymq2017/entitysam) (CVPR2025) performs automatic segmentation and tracking of novel entities without requiring user specified prompts.

OK thank you~~
